nomad/heartbeat.go

```diff
@@ -142,15 +145,37 @@ func (h *nodeHeartbeater) invalidateHeartbeat(id string) {
 	}
 	h.heartbeatTimersLock.Unlock()
 
-	// Do not invalidate the node since we are not the leader. This check avoids
+	// Do not invalidate the node if we are no longer the leader. This check avoids
 	// the race in which leadership is lost but a timer is created on this
 	// server since it was servicing an RPC during a leadership loss.
-	if !h.IsLeader() {
-		h.logger.Debug("ignoring node TTL since this server is not the leader", "node_id", id)
+	if future := h.raft.VerifyLeader(); future.Error() != nil {
+		h.logger.Debug("ignoring expired node TTL since this server is not the leader", "node_id", id)
+		return
+	}
+
+	// Lookup the node and see if a node update got sent while we were confirming
+	// out status as leader.
+	snap, err := h.fsm.State().Snapshot()
+	if err != nil {
+		h.logger.Error("failed to create fsm snapshot to lookup node TTL status", "node_id", id, "error", err)
+		return
+	}
+	node, err := snap.NodeByID(memdb.NewWatchSet(), id)
+	if err != nil {
+		h.logger.Error("failed to lookup node for TTL expiration status", "node_id", id, "error", err)
+		return
+	}
+
+	// Check if a recent node update came in after the deadline. If so, we just
+	// reset the heartbeat timer.
+	latestUpdate := time.Unix(node.StatusUpdatedAt, 0)
+	if latestUpdate.After(fencepost) {
+		_, _ = h.resetHeartbeatTimer(id)
+		h.logger.Debug("reset node TTL because a newer status update was found", "node_id", id)
 		return
 	}
 
-	h.logger.Warn("node TTL expired", "node_id", id)
+	h.logger.Warn("setting node status down due to expired TTL", "node_id", id, "fencepost", fencepost, "latest_update", latestUpdate)
```

Comment:
A nitpick: we might not want to call it a "fencepost" in the logs. It's obvious to us reading the code but if you're an operator what does that mean? Maybe something like "missed_at"?